### PR TITLE
Implement shape-inference caching for codegen LazyNativeFunctions

### PR DESCRIPTION
**Pull request by wconstab**
RUN_TORCHBENCH: ALL
TORCHBENCH_BRANCH: wconstab/ltc

- Refactor to construct node hashes in generated LazyNativeFunctions
- Provide backward-compatible TsNode ctors to support legacy
  (non-codegen) IR classes
- Add codegen support for shape cache

Example generated code (addcdiv)
```
// TODO(alanwaketan): Public members don't need to have _ suffix.
class Addcdiv : public torch::lazy::TsNode {
 public:
  Addcdiv(const torch::lazy::Value& self, const torch::lazy::Value& tensor1, const torch::lazy::Value& tensor2, const at::Scalar& value, torch::lazy::hash_t node_hash, torch::lazy::hash_t dag_hash, std::vector<Shape>&& shapes)
      : torch::lazy::TsNode(torch::lazy::OpKind(at::aten::addcdiv),
              {self, tensor1, tensor2}, std::move(shapes),
              /* num_outputs */ 1,
              node_hash,
              dag_hash),

        value_(value)

  {
    
  }

  std::string ToString() const override {
    std::stringstream ss;
    ss << TsNode::ToString();
    ss << ", value=" << value_;
    return ss.str();
  }

  torch::lazy::TSOpVector Lower(std::shared_ptr<torch::jit::GraphFunction> function,
                   torch::lazy::TSLoweringContext* loctx) const override {
        std::vector<torch::jit::NamedValue> arguments;
    std::vector<torch::jit::NamedValue> kwarguments;
    arguments.reserve(3);
    kwarguments.reserve(1);
    size_t i = 0;
    arguments.emplace_back(loctx->GetOutputOp(operand(i++)));
    arguments.emplace_back(loctx->GetOutputOp(operand(i++)));
    arguments.emplace_back(loctx->GetOutputOp(operand(i++)));
    kwarguments.emplace_back("value", value_);
    torch::lazy::TSOpVector addcdiv_out = torch::lazy::LowerTSBuiltin(function, op().op, arguments, kwarguments);
    CHECK_EQ(addcdiv_out.size(), 1);

    // TODO: need to call GenerateClone sometimes? Or else return LowerBuiltIn() directly
    return addcdiv_out;

  }

  at::Scalar value_;
  

};

    // TODO(alanwaketan): Quite a lot inefficient copy-by-value there. Let's optimize it.
    at::Tensor LazyNativeFunctions::addcdiv(const at::Tensor & self, const at::Tensor & tensor1, const at::Tensor & tensor2, const at::Scalar & value) {
        LTC_FN_COUNTER("lazy::");
        auto device = bridge::GetBackendDevice(self, tensor1, tensor2);
        LazyTensor lazy_self = GetLtcTensorOrCreateForWrappedNumber(self, *device);
    LazyTensor lazy_tensor1 = GetLtcTensorOrCreateForWrappedNumber(tensor1, *device);
    LazyTensor lazy_tensor2 = GetLtcTensorOrCreateForWrappedNumber(tensor2, *device);
        torch::lazy::hash_t node_hash = torch::lazy::MHash(static_cast<uint32_t>(at::aten::addcdiv), value);
        torch::lazy::hash_t dag_hash = torch::lazy::OperandHashes({lazy_self.GetIrValue(), lazy_tensor1.GetIrValue(), lazy_tensor2.GetIrValue()}, node_hash);
        
        auto shape_cache = torch::lazy::GetShapeCache();
        auto cached_shapes = shape_cache->Get(dag_hash);
        if (cached_shapes == nullptr) {
            auto out_meta = at::meta::addcdiv(self, tensor1, tensor2, value);
            cached_shapes = shape_cache->Add(dag_hash, std::make_shared<std::vector<Shape>>(std::initializer_list<Shape>{Shape(out_meta.scalar_type(), out_meta.sizes().vec())}));

        }
        auto& shapes = *cached_shapes;
        TORCH_INTERNAL_ASSERT(shapes.size() == 1);

        auto node = torch::lazy::MakeNode<ir::ops::Addcdiv>(lazy_self.GetIrValue(),
                              lazy_tensor1.GetIrValue(),
                              lazy_tensor2.GetIrValue(),
                              value,
                              node_hash,
                              dag_hash,
                                                                                      std::move(shapes));
        auto result = CreateAtenFromLtcTensor(LazyTensor::Create(node, lazy_self.GetDevice()));
        return result;
    };


```

Timing results:
note: ADAM learning-rate hack is to modify adam optimizer in _functional.py to pass a non-changing constant for the lastparam of addcdiv_ to prevent recompilation on every step due to not handling scalars well.

Baseline + ADAM learning-rate hack:
hf_Bert                        cuda eval  trace overheads      overhead: 0.204 pvalue: 9.43e-08
hf_Bert                        cuda eval  amortized 10x        speedup:  0.958 pvalue: 4.05e-04
hf_Bert                        cuda eval  unamortized          speedup:  0.756 pvalue: 2.14e-07
hf_Bert                        cuda train trace overheads      overhead: 0.162 pvalue: 4.89e-10
hf_Bert                        cuda train amortized 10x        speedup:  0.981 pvalue: 3.83e-04
hf_Bert                        cuda train unamortized          speedup:  0.841 pvalue: 4.65e-07

This PR + ADAM learning-rate hack
hf_Bert                        cuda eval  trace overheads      overhead: 0.166 pvalue: 3.98e-09
hf_Bert                        cuda eval  amortized 10x        speedup:  0.963 pvalue: 1.56e-05
hf_Bert                        cuda eval  unamortized          speedup:  0.782 pvalue: 4.29e-07
hf_Bert                        cuda train trace overheads      overhead: 0.171 pvalue: 1.55e-07
hf_Bert                        cuda train amortized 10x        speedup:  0.981 pvalue: 2.85e-04
hf_Bert                        cuda train unamortized          speedup:  0.853 pvalue: 1.87e-07

image
image

This PR doesn't make a significant difference to bert train/eval time, even though it appears to reduce the 'overhead' considerably.  This suggests the remaining gap between eager and lazy may either be 
 - slower execution time for 'optimized/compiled' gpu code
 - other synchronization issues that prevent fully overlapping tracing with execution

Looking at nsys kernel traces suggests that the time spent running GPU kernels under lazy/nvfuser is faster by a hair (1.480 / 1.474 = 1.0041), meanwhile the time waiting before launching the first gpu kernel is almost unchanged (30.8 ms baseline vs 30.5ms with cache).

The eager cuda sync-sync bert step time is about 149ms
30 / 149 = 20% 'trace overhead'.  We measured 20% trace overhead for the baseline, but 16% trace overhead for the shape cache.  
16% of 149 would be 24ms, so somehow there is a 6ms gap between how we measure 'trace overhead' and how long it actually takes to launch gpu kernels .

### CODE REVIEW

File: tools/codegen/dest/lazy_ir.py [hunk 1]
```diff
@@ -39,10 +54,41 @@ def node_ctor_inputs(func: LazyIrSchema) -> str:
...
+    else:
+        shape_sig = ComputeShapeSignature(func)
+        shapes_body = f"""
+            cached_shapes = shape_cache->Add(dag_hash, std::make_shared<std::vector<Shape>>({shape_sig.shape_call}));
+"""    
+    
+    shapes_str = f"""
+        auto shape_cache = torch::lazy::GetShapeCache();
+        auto cached_shapes = shape_cache->Get(dag_hash);
+        if (cached_shapes == nullptr) {{{shapes_body}
```
Comment (desertfire) [inline tools/codegen/dest/lazy_ir.py:85]: Nit: the formatting is a little strange here.

File: torch/csrc/lazy/ts_backend/ts_node.cpp [hunk 2]
```diff
@@ -40,15 +39,57 @@ hash_t OperandHashes(const OpList& operands, const hash_t& seed) {
...
+
+TsNode::TsNode(OpKind op, OpList operands,
+               const std::function<Shape()>& shape_fn,
+               size_t num_outputs, hash_t hash_seed)
+    : TsNode(op, operands, std::vector<Shape>{}, num_outputs, 
            /* node_hash */ HashCombine(op.hash(), hash_seed),
            /* dag_hash */
-           OperandHashes(operands, HashCombine(op.hash(), hash_seed))),
+           OperandHashes(operands, HashCombine(op.hash(), hash_seed))) {
+  // shapes_.push_back(GetOpShape(shape_fn));
```
Comment (desertfire) [inline torch/csrc/lazy/ts_backend/ts_node.cpp:61]: Any reason why we are keeping this as a comment here?

**Comment by desertfire** [comment]
It will be helpful to post a new generated code example here.

File: tools/codegen/dest/lazy_ir.py (hunk 1 above)
Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:85]: this was at least intentional, 

shapes_body starts out with a newline and then always has the same number of indents

leaving the }} to the next line makes it show up correctly.

**Review by desertfire — APPROVED** [review]

File: tools/codegen/dest/lazy_ir.py [hunk 3]
```diff
@@ -58,6 +73,37 @@ def aten_symbol(schema: LazyIrSchema) -> str:
         return f'c10::Symbol::fromQualString("aten::{schema.aten_name}")'
     return f'at::aten::{schema.aten_name}'
 
+def cached_shape_inference(func: NativeFunction, schema: LazyIrSchema,
+                           returns_length: int, all_types: List[NamedCType]) -> str:
+    # call the meta kernel if it exists, to compute output shape/dtype for our IR
+    if func.structured or func.structured_delegate is not None:
```
Comment (Krovatkin) [inline tools/codegen/dest/lazy_ir.py:79]: we are effectively caching on a rank here?

File: tools/codegen/dest/lazy_ir.py [hunk 4]
```diff
@@ -58,6 +73,37 @@ def aten_symbol(schema: LazyIrSchema) -> str:
         return f'c10::Symbol::fromQualString("aten::{schema.aten_name}")'
     return f'at::aten::{schema.aten_name}'
 
+def cached_shape_inference(func: NativeFunction, schema: LazyIrSchema,
```
Comment (Krovatkin) [inline tools/codegen/dest/lazy_ir.py:76]: ~~This code guarded by `shape_cache->Get(dag_hash)` somehow? I.e. we wouldn't call into meta kernels if we already have the shapes for a given `dag_hash`?~~

found it! 

```
        auto cached_shapes = shape_cache->Get(dag_hash);
        if (cached_shapes == nullptr) {{{shapes_body}
```

File: tools/codegen/dest/lazy_ir.py [hunk 5]
```diff
@@ -58,6 +73,37 @@ def aten_symbol(schema: LazyIrSchema) -> str:
...
+                           returns_length: int, all_types: List[NamedCType]) -> str:
+    # call the meta kernel if it exists, to compute output shape/dtype for our IR
+    if func.structured or func.structured_delegate is not None:
+        meta_out = """cached_shapes = shape_cache->Add(dag_hash, std::make_shared<std::vector<Shape>>(std::initializer_list<Shape>{Shape(out_meta.scalar_type(), out_meta.sizes().vec())}));"""
+        if returns_length > 1:
+            def this_shape(i: int) -> str:
+                return f"Shape(std::get<{i}>(out_meta).scalar_type(), std::get<{i}>(out_meta).sizes().vec())"
```
Comment (Krovatkin) [inline tools/codegen/dest/lazy_ir.py:83]: why are we using full shapes here compared to only using a rank above?

**Review by Krovatkin — APPROVED** [review]
:shipit:

looks gut.

File: tools/codegen/dest/lazy_ir.py (hunk 3 above)
Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:79]: we are caching on dag-hash, which currently includes shape but after your change would only include rank.  but more to the point, we're caching on the same thing we use for lazy graph cache, so it should make this shape inference caching always consistent with non-cached verison, right?

File: tools/codegen/dest/lazy_ir.py (hunk 5 above)
Comment (wconstab) [reply tools/codegen/dest/lazy_ir.py:83]: i'm not sure i understand what you mean.

**Comment by wconstab** [comment]
Oops, didn't notice this is failing 3 tests:
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] AtenLtcTsTensorTest.TestClampMax
[  FAILED  ] AtenLtcTsTensorTest.TestCelu
[  FAILED  ] AtenLtcTsTensorTest.TestCeluInPlace

not ready to merge
cc @desertfire

**Pull request closed without merging**